File: packages/node/src/managed_service_scanner.ts
// PRIVACY POLICY: Scans for presence of env vars by NAME only.
// Values are passed through techFromUrl() which strips credentials/hosts.
// Raw values are never stored or shipped.

import { techFromUrl } from "./sanitizer"

interface Discovery {
  type: string
  tech: string
  source: string
  variable_name: string
}

const DB_ENV_VARS    = ["DATABASE_URL", "DB_URL", "POSTGRES_URL", "MYSQL_URL", "MONGODB_URI"]
const CACHE_ENV_VARS = ["REDIS_URL", "REDIS_TLS_URL", "UPSTASH_REDIS_REST_URL"]
const QUEUE_ENV_VARS = ["RABBITMQ_URL", "AMQP_URL", "CLOUDAMQP_URL"]

export function scanManagedServices(): Discovery[] {
  const discoveries: Discovery[] = []
  const seen = new Set<string>()

  for (const varName of DB_ENV_VARS) {
    const val = process.env[varName]
    if (!val) continue
    const tech = techFromUrl(val)
    if (!tech) continue
    const key = `database:${tech}`
    if (seen.has(key)) continue
    seen.add(key)
    discoveries.push({ type: "database", tech, source: "env_detection", variable_name: varName })
  }

  for (const varName of CACHE_ENV_VARS) {
    if (!(varName in process.env) || !process.env[varName]) continue
    const key = "cache:redis"
    if (seen.has(key)) continue
    seen.add(key)
    discoveries.push({ type: "cache", tech: "redis", source: "env_detection", variable_name: varName })
  }

  for (const varName of QUEUE_ENV_VARS) {
    const val = process.env[varName]
    if (!val) continue
    const tech = techFromUrl(val)
    if (!tech) continue
    const key = `queue:${tech}`
    if (seen.has(key)) continue
    seen.add(key)
    discoveries.push({ type: "queue", tech, source: "env_detection", variable_name: varName })
  }

  return discoveries
}

File: packages/ruby/lib/nurse_andrea/managed_service_scanner.rb
# PRIVACY: This file reads environment variable values to derive
# component metadata. Raw values never leave this process. Only
# derived metadata (type, tech, provider) is transmitted.
# See DATA_PRIVACY_POLICY.rb for the full policy.

module NurseAndrea
  module ManagedServiceScanner
    SCAN_MAP = {
      "DATABASE_URL"      => "database",
      "DATABASE_PRIVATE_URL" => "database",
      "REDIS_URL"         => "cache",
      "REDIS_PRIVATE_URL" => "cache",
      "RABBITMQ_URL"      => "queue",
      "CLOUDAMQP_URL"     => "queue",
      "MONGODB_URI"       => "database",
      "MONGO_URL"         => "database",
      "ELASTICSEARCH_URL" => "search",
      "KAFKA_BROKERS"     => "queue"
    }.freeze

    def self.scan
      # Skip env-based discovery entirely when the SDK is loaded inside
      # NurseAndrea itself — every URL we'd find belongs to the platform's
      # own infrastructure, not a customer component.
      return [] if NurseAndrea::SelfFilter.platform_self?

      discoveries = []

      SCAN_MAP.each do |var_name, component_type|
        url = ENV[var_name]
        next if url.nil? || url.strip.empty?
        next if NurseAndrea::SelfFilter.host_matches?(url)

        tech     = Sanitizer.extract_tech(url)
        provider = Sanitizer.extract_provider(url)

        raw = {
          type: component_type,
          tech: tech,
          provider: provider,
          source: "env_detection",
          variable_name: var_name
        }

        discoveries << Sanitizer.sanitize_discovery(raw)
      end

      discoveries.uniq { |d| [ d[:type], d[:tech], d[:provider] ] }
    end
  end
end
